x-pack/plugins/security/server/audit/audit_events.ts

```diff
@@ -0,0 +1,245 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { KibanaRequest } from 'src/core/server';
+import { AuthenticationResult } from '../authentication/authentication_result';
+
+/**
+ * Audit event schema using ECS format.
+ * https://www.elastic.co/guide/en/ecs/1.5/index.html
+ * @public
+ */
+export interface AuditEvent {
+  /**
+   * Human readable message describing action, outcome and user.
+   *
+   * @example
+   * User [jdoe] logged in using basic provider [name=basic1]
+   */
+  message: string;
+  event: {
+    action: string;
+    category?: EventCategory;
+    type?: EventType;
+    outcome?: EventOutcome;
+    module?: string;
+    dataset?: string;
+  };
+  user?: {
+    name: string;
+    email?: string;
+    full_name?: string;
+    hash?: string;
+    roles?: readonly string[];
+  };
+  kibana?: {
+    /**
+     * Current space id of the request.
+     */
+    space_id?: string;
```

Comment:
I noticed `saved_object` (below) does **not** have a `space_id` - is the intention that the referenced saved object is available scoped to the `space_id` specified here - wondering about SO's available in multiple spaces.

We have an _potential_ interesting "opposite" problem here for alerting - if we ever needed to generate an audit record when an alert or action was running, there isn't really a "current" space in effect, alerts and actions run via task manager.  But we could assume the `space_id` of the space associated with the SO passed to those task manager runners. 

In the end, I'm guessing we won't have any audit records that need to be run in the background like that, I'd think most audit needs would be based on explicit HTTP requests instead.

Reply:
The reason we removed the `space_id` field from the `saved_object` is that its meaning was ambiguous. It wasn't clear whether it referred to all the spaces the object exists in or the spaces that were added to, or removed from the saved object as part of a `saved_object_add_to_spaces` / `saved_object_delete_from_spaces` event. 

Instead, for those events, we've added an explicit `add_to_spaces` / `delete_from_spaces` param to make its meaning less ambiguous:

```json
{
  "event": {
    "action": "saved_object_ add_to_spaces"
  },
  "kibana": {
    "space_id": "default",
    "saved_object": { "type": "dashboard", "id": "123" },
    "add_to_spaces": ["space2"]
  }
}
```

The AuditEvent schema is left intentionally open for use cases like this so we could add an event specific `associated_spaces` field if necessary. Alternatively, we could also override the existing `space_id` field with the associated space of the alert. This would only make sense if an alert is always generated in a single space though, even when a saved object is shared across multiple spaces.